### ISSUE TITLE
[Fatal] FEServer (v1.4.0) : Desync'd HP

### ISSUE THREAD

**Issue by SapphireTactician** [eliatlarge/FEMultiPlayer-V2#89]
Issue type: `problem`

#### Description:

So something similar to [this](https://github.com/rayrobdod/FEMultiPlayer-V2/issues/3) actually happened in the wild on the night of the tournament during one of my friendlies with **Pwntagonist**, or so I believe, resulting in FEServer catching fire. I believe this is similar to @rayrobdod own findings because when it crashed, **Pwntagonist** said in Discord something along the lines of:

>"Yeah, I probably hit the button too fast"

, upon showing him the log as proof that I didn't rage quit (*especially because I was confident the battle would end in a stalement and I don't rage quit*). This is a bug @rayrobdod suspected to be in upstream, and I can now confirm.

I'm pretty sure my PC in this instance was the point of failure because when having to run both the client, the server, AND Discord most likely choked my CPU (*__due to #42 yet again...__ this bug will be the death of me! Btw, any luck with my new snapshots from Pull #87 yet? @rayrobdod*) and thus creating a scenario where it was possible that my opponent could've probably initiated a new combat sequence before a previous one could finish on my end. Which @rayrobdod suspects to be the cause when he was experimenting with clamped framerates via #87 .
  
However, as per [this post](https://www.reddit.com/r/FEMP/comments/4ji7n2/thoughts_on_the_tourney_among_other_things/), I'm not sure if this may be a false positive due to an experiment @eliatlarge performed in order to make FEServer a truly standalone application instead of a headless client without anyone's knowledge which may or may not have been reversed - I'll most likely be testing v1.4.1 tonight with @JustaLevelZero .

-------

#### Examples:

    java.lang.RuntimeException: Desynced HP: Oswin
        at net.fe.network.message.EndTurn.checkHp(EndTurn.java:42)
        at net.fe.overworldStage.OverworldStage.beginStep(OverworldStage.java:275)
        at net.fe.network.FEServer.loop(FEServer.java:347)
        at net.fe.network.FEServer$4$2.run(FEServer.java:263)

Attached are the server error log and log of the match itself:

[HPDesyncFatal_logs.zip](https://github.com/eliatlarge/FEMultiPlayer-V2/files/268976/HPDesyncFatal_logs.zip)

**Comment by eliatlarge** [comment]
See, I removed several res files from the game, possibly including a stat file. Ant is putting duplicates of every file in both its own dir and the main dir, so if I deleted one, the server may have been using the other. If you could, please try using 1.4_fix1 and tell me if it pops up again.

I'm very suspicious this is the case especially because the logs for desync come fro a server log, and the code has an exception specifically for server:

```
public void checkHp(boolean server){
		if(server){
			for(UnitIdentifier u: unitHp.keySet()){
				if(FEServer.getUnit(u).getHp() != unitHp.get(u)){
					throw new RuntimeException("Desynced HP: " + u.name);
				}
			}
		} else {
			for(UnitIdentifier u: unitHp.keySet()){
				if(FEMultiplayer.getUnit(u).getHp() != unitHp.get(u)){
					throw new RuntimeException("Desynced HP: " + u.name);
				}
			}
		}
	}
```

this is under EndTurn.java. I wouldn't be surprised if the majority of false positives are a result of my fooling around, but at the same time, I'd also like to fix ant so that the filesize isn't as big for players with worse connections.

**Comment by SapphireTactician** [comment]
Instead of calling it `_fix1` or whatever, to make versioning easier, I'm sure you know the proper numbering scheme, right? `[major].[minor].[fix]` or `[major].[minor - such as new features].[enhancement - such as small tweaks, balancing, etc.].[fix - bug and design flaw fixes]`

**Comment by eliatlarge** [comment]
I really just use fix# for immediate patches, rather than new versions. fix1, imo, tells unexperienced peeps (the majority of the fanbase) what the new version is/means.

**Comment by SapphireTactician** [comment]
**Pwntagonist** commented on pressing "A" (*by which I'm assuming he means "Z" due to it being mapped as the GBA "A" button*) to soon / quickly. I'm led to believe this may be related to #92 or #93

**Comment by SapphireTactician** [comment]
Closing as `Fixed` as per #102 

Great job!
